client/canonical_eventlog.go

```diff
@@ -0,0 +1,266 @@
+
+	buf[0] = tlv.Tag
+	binary.BigEndian.PutUint32(buf[tlvTagFieldLength:], tlv.Length)
+	copy(buf[tlvTagFieldLength+tlvLengthFieldLength:], tlv.Value)
+
+	return buf
+}
+
+// TLVUnmarshal reads and parse the first TLV from the bytes buffer.
+func TLVUnmarshal(buf *bytes.Buffer) (TLV, error) {
+	var tlv TLV
+	var err error
+
+	// read tag
+	tlv.Tag, err = buf.ReadByte()
+	if err != nil {
+		return TLV{}, err
+	}
+
+	// read length
+	b := make([]byte, tlvLengthFieldLength)
+	l, err := buf.Read(b)
+
+	if err != nil || l < tlvLengthFieldLength {
+		return TLV{}, io.EOF
+	}
+	tlv.Length = binary.BigEndian.Uint32(b)
+
+	// read value
+	tlv.Value = make([]byte, tlv.Length)
+	l, err = buf.Read(tlv.Value)
+	if err != nil || l < int(tlv.Length) {
+		return TLV{}, io.EOF
+	}
+	return tlv, nil
+}
+
+// CELR represents a Canoical Eventlog Record
+type CELR struct {
+	RecNum  TLV
+	PCR     TLV
+	Digests TLV
+	Content TLV
+}
+
+// CELREventData is a interface for the content in CELR.
+type CELREventData interface {
+	GenerateDigest(crypto.Hash) ([]byte, error)
+	GetTLV() TLV
+}
+
+// CEL represents a Canonical Eventlog, which contains a list of CELR
+type CEL struct {
+	Records []CELR
+}
+
+// AppendEvent appends a new record to the CEL.
+func (c *CEL) AppendEvent(tpm io.ReadWriteCloser, pcr int, hashAlgos []crypto.Hash, payload CELREventData) error {
+	digestsMap := make(map[tpm2.Algorithm][]byte)
+	for _, hashAlgo := range hashAlgos {
+		digest, err := payload.GenerateDigest(hashAlgo)
+		if err != nil {
+			return err
+		}
+		digestsMap[hashAlgoMapping[hashAlgo]] = digest
+
+		// TODO: extend the digest to TPM PCR
+	}
+
+	celr, err := createCELR(uint64(len(c.Records)), uint8(pcr), digestsMap, payload)
+	if err != nil {
+		return err
+	}
+
+	c.Records = append(c.Records, celr)
+	return nil
+}
+
+func createRecNumField(recNum uint64) TLV {
+	payload := make([]byte, recNumValueLength)
+	binary.BigEndian.PutUint64(payload, recNum)
+	return TLV{recNumTagValue, recNumValueLength, payload}
+}
+
+// UnmarshalRecNum takes in a TLV with its tag equals to the recnum tag value (0), and
+// return its record number.
+func UnmarshalRecNum(tlv TLV) (uint64, error) {
+	if tlv.Tag != recNumTagValue {
+		return 0, fmt.Errorf("tag of the TLV indicates it is not a recnum field")
+	}
+	return binary.BigEndian.Uint64(tlv.Value), nil
+}
+
+func createPCRField(pcrNum uint8) TLV {
+	return TLV{pcrTagValue, pcrValueLength, []byte{pcrNum}}
+}
+
+// UnmarshalPCR takes in a TLV with its tag equals to the PCR tag value (3), and
+// return its PCR number.
+func UnmarshalPCR(tlv TLV) (pcrNum uint8, err error) {
+	if tlv.Tag != pcrTagValue {
+		return 0, fmt.Errorf("tag of the TLV indicates it is not a PCR field")
+	}
+	return tlv.Value[0], nil
+}
+
+func createDigestField(digestMap map[tpm2.Algorithm][]byte) TLV {
+	var buf bytes.Buffer
+
+	for hashAlgo, hash := range digestMap {
+		singledigest := TLV{uint8(hashAlgo), uint32(len(hash)), hash}
```

Comment:
let's check that the digest size matches the algorithm's size

Reply:
Done